### PR TITLE
Add `AsyncSeek` implementation for `BufReader` and `BufWriter`

### LINKED ISSUE #2291: tokio-io: No `AsyncSeek` implementation for `BufReader` and `BufWriter`
## Version

[`v0.2.13`](https://github.com/tokio-rs/tokio/releases/tag/tokio-0.2.13)

## Description

Rust's standard library provides `Seek` implementation of `BufReader` and `BufWriter`. Also, `futures` crate provides `AsyncSeek` implementation for `BufReader` and `BufWriter`. I think, it'd be a good idea to implement [`tokio::io::AsyncSeek`](https://docs.rs/tokio/0.2.13/tokio/io/trait.AsyncSeek.html) for [`tokio::io::BufReader`](https://docs.rs/tokio/0.2.13/tokio/io/struct.BufReader.html) and [`tokio::io::BufWriter`](https://docs.rs/tokio/0.2.13/tokio/io/struct.BufWriter.html).

### PR DESCRIPTION
**Pull request by devashishdxt**
## Motivation

Fixes #2291.

## Solution

Implementation is based on current `AsyncSeek` implementation for `BufReader` and `BufWriter` in `futures` crate.

### CODE REVIEW

File: tokio/src/io/util/buf_reader.rs [hunk 1]
```diff
@@ -153,6 +153,67 @@ impl<R: AsyncRead> AsyncBufRead for BufReader<R> {
...
+                // seek backwards by our remainder, and then by the offset
+                ready!(self
+                    .as_mut()
+                    .get_pin_mut()
+                    .start_seek(cx, SeekFrom::Current(-remainder)))?;
+                self.as_mut().discard_buffer();
+                ready!(self
+                    .as_mut()
+                    .get_pin_mut()
+                    .start_seek(cx, SeekFrom::Current(n)))?;
```
Comment (Darksonn) [inline tokio/src/io/util/buf_reader.rs]: This is correct, but it is non-obvious. I would like to see a comment that explains why the `discard_buffer()` call means we don't end up in this branch again, even if the second seek returns pending.

**Comment by Darksonn** [comment]
Hey @devashishdxt. Are you interested in getting this merged?

**Comment by devashishdxt** [comment]
> Hey @devashishdxt. Are you interested in getting this merged?

Yes. I was a little busy these past few days. I’ll try to add some test cases over this weekend. I too find it difficult to understand the login in code (I just copy-pasted this code from `futures` crate). So, I may need some help to do it.

**Comment by Darksonn** [comment]
Sure, if you need any help, feel free to jump on [our discord server][1].

[1]: https://discord.gg/tokio

**Comment by Darksonn** [comment]
Hey @devashishdxt, I would love to have this merged. If you are interested in getting it across the finish line, that would be great. Alternatively I can add the relevant comment to the PR and have it merged.

**Comment by carllerche** [comment]
Closing as this has gone stale.

**Pull request closed without merging**